File: 2026/aed-1/cormen/2.3/merge_sort.cpp
#include <iostream>

void merge(int *A, int p, int q, int r) {
  int n_L = q - p + 1;
  int n_R = r - q;
  int L[n_L];
  int R[n_R];

  for (int i = 0; i < n_L; i++) {
    L[i] = A[p + i];
  }

  for (int i = 0; i < n_R; i++) {
    R[i] = A[q + i + 1];
  }

  int i = 0;
  int j = 0;
  int k = p;

  while (i < n_L && j < n_R) {
    if (L[i] <= R[j]) {
      A[k++] = L[i++];
    } else {
      A[k++] = R[j++];
    }
  }

  while (i < n_L) {
    A[k++] = L[i++];
  }

  while (j < n_R) {
    A[k++] = R[j++];
  }
}

void merge_sort(int *A, int p, int r) {
  if (p >= r) {
    return;
  }

  int q = ((p + r) / 2);

  merge_sort(A, p, q);
  merge_sort(A, q + 1, r);
  merge(A, p, q, r);
}

int main() {
  int A[5] = {4, 3, 1, 6, 10};
  int n = sizeof(A) / sizeof(int);
  int p = 0;
  int r = 4;

  std::cout << "Unsorted." << '\n';

  for (int i = 0; i < n; i++) {
    std::cout << A[i] << '\n';
  }

  merge_sort(A, p, r);

  std::cout << "Sorted." << '\n';

  for (int i = 0; i < n; i++) {
    std::cout << A[i] << '\n';
  }

  return 0;
}

File: 2026/aed-1/cormen/2.3/merge_sort.rs
fn merge(a: &mut [i32], p: usize, q: usize, r: usize) {
    let n_l = q - p + 1;
    let n_r = r - q;
    let mut l = vec![0; n_l];
    let mut r = vec![0; n_r];

    for i in 0..n_l {
        l[i] = a[p + i];
    }

    for i in 0..n_r {
        r[i] = a[q + i + 1];
    }

    let mut i = 0;
    let mut j = 0;
    let mut k = p;

    while i < n_l && j < n_r {
        if l[i] <= r[j] {
            a[k] = l[i];
            i = i + 1;
            k = k + 1;
        } else {
            a[k] = r[j];
            j = j + 1;
            k = k + 1;
        }
    }

    while i < n_l {
        a[k] = l[i];
        i = i + 1;
        k = k + 1;
    }

    while j < n_r {
        a[k] = r[j];
        j = j + 1;
        k = k + 1;
    }
}

fn merge_sort(a: &mut [i32], p: usize, r: usize) {
    if p >= r {
        return;
    }

    let q = (p + r) / 2;

    merge_sort(a, p, q);
    merge_sort(a, q + 1, r);
    merge(a, p, q, r);
}

fn main() {
    let mut a = [4, 3, 1, 6, 10];
    let p = 0;
    let r = 4;

    println!("Unsorted.");

    for i in &a {
        println!("{}", i);
    }

    merge_sort(&mut a, p, r);

    println!("Sorted.");

    for i in &a {
        println!("{}", i);
    }
}
